File: Java/com/leetcode/pathSumII/PathSumII.java
package com.leetcode.pathSumII;

import java.util.ArrayList;
import java.util.List;

import com.leetcode.dataStructures.TreeNode;

public class PathSumII {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        ArrayList<ArrayList<Integer>> pathes = new ArrayList<ArrayList<Integer>>();
        
        if (root == null) {
            return (List)pathes;
        }
        
        DFS(root, 0, sum, pathes, new ArrayList<Integer>());
        
        return (List)pathes;
    }
    
    private void DFS(TreeNode root, int currentSum, int sum, ArrayList<ArrayList<Integer>> pathes, ArrayList<Integer> path) {
        
        currentSum += root.val;
        if (currentSum == sum && root.left == null && root.right == null) {
            path.add(root.val);
            pathes.add(path);
            return;
        }
        

        if (root.left != null) {
            ArrayList<Integer> clonePath = (ArrayList<Integer>)path.clone();
            clonePath.add(root.val);
            DFS(root.left, currentSum, sum, pathes, clonePath);
        }
        
        if (root.right != null) {
            ArrayList<Integer> clonePath = (ArrayList<Integer>)path.clone();
            clonePath.add(root.val);
            DFS(root.right, currentSum, sum, pathes, clonePath);
        }
    }
}

File: Python/path_sum_ii/path_sum_ii.py
class Solution(object):
    def pathSum(self, root, sum):
        results = []
        self.helper(root, [], results, sum)
        return results

    def helper(self, root, curr, results, s):
        if root is None:
            return
        s -= root.val
        new_curr = curr + [root.val]
        if s == 0 and root.left is None and root.right is None:
            results.append(new_curr)
            return

        self.helper(root.left, new_curr, results, s)
        self.helper(root.right, new_curr, results, s)
